### PR TITLE
Data loading

### LINKED ISSUE #28: Error when selecting size
Error occurs when size value is set to certain columns b/c they contain negative values

### LINKED ISSUE #30: Create Landing Page
The landing page, located in `application/templates/index.html` needs to be created and styled.  Currently there are just 2 anchors in there, but both other anchors will need to be added, as well as a ui design for the landing page.  CSS styling can be done inline, but putting it in another file would be preferable.  Having the css in a separate file requires adding asset bundles to the flask routes, as does adding js.  See #31

### LINKED ISSUE #31: Add Asset Bundle handling
In order to have JS or CSS styling for the pages, asset bundle handling needs to be added in `/application/__init__.py`. #30 will be easier once this is implemented.

### LINKED ISSUE #40: Subanalysis of dataset
Allow some type of analysis on a subset of the data. User should be able to select which points to include in the main plot, then either reload with only those points or open them in a separate plot.

### LINKED ISSUE #45: Add html function/seperate file
**Is your feature request related to a problem? Please describe.**
Raw html currently is just stored in strings in files such as `wsgi.py`, having a function to load that would be preferable.
**Describe the solution you'd like**
An function which gets html from a separate file is needed, so that there isn't just raw html in various files.  Alternatively, using Jinja2 or another templating system with flask might be simpler.

### LINKED ISSUE #49: Data format between Tesla-ver and Loclust
Currently, Tesla-ver parses a group of loclust trajectory files into a single dataframe that's then used to render the graph.

Providing interoperability with loclust would be useful, as is shifting it to apply to a single data file with multiple trajectories (a truly 3D or greater dataset, compared to the 2D data files it's being tested with right now).

Mostly this issue exists to document and open a discussion on which data tesla-ver is supposed to be used, and how the data is then stored and handled internally.

### PR DESCRIPTION
**Pull request by alexanderkyim**
---
name: Data Loading
about: Added a dedicated data uploading screen and redis store between the two figures 

---

## Issues Addressed
Closes #49, Closes #40, Closes #28, Closes #45, Closes #30, Closes #31 

## What has changed
Include each file that was changed, how it was changed (in broad terms), and why it was changed.
Example:
    tesla-ver/bubble_chart.py
       Changed to load data from redis, rather than parse it locally in this screen
    tesla-ver/bubble_chart_layout.py
       Removed data uploading components
    tesla-ver/data_uploading.py
        Added, now handles uploading and pushing data to redis
    tesla-ver/data_uploading_layout.py
        Provides the layout for the data uploading screen

## Checklist of pre-requisites
All of these need to be checked in order to request a review on the Pull Request
-   [X] Does the code run?  
-   [X] Does the code follow the repository style?  
-   [X] Is the code tested?  
-   [X] Has Black been run on all the python files in the repository (with line length 120 `black -l 120 *`)?

## Additional notes

# Enhancement

## How to use the feature
Build and launch the project with `docker-compose up --build`, and then go to the data uploading screen.  Upload data there, select the identifying column, year column, and data columns to graph, as well as any rows that you wish to graph (if no rows are selected, all rows will be graphed)

## If the user interface was modified include screenshots here

Screen Shot 2020-07-11 at 01 53 40

# Design Change

## How has the design changed?
The project now runs inside a docker container

## Why that change was necessary?
Without a redis memory store, it would have relied on workarounds to change screens for a data uploading screen.  Running this as a docker compose project allows a redis datastore and the project to be started right away and linked immediately.

### CODE REVIEW

File: tesla_ver/bubble_chart/bubble_chart.py [hunk 1]
```diff
@@ -47,6 +47,11 @@ def extract_mdata(df, x_column_name):
                 "time_max": max(x_vals),
                 "x_vals": x_vals,
                 "data_cols": [col for col in df.columns.values.tolist() if col not in ["X", "Year", "Subject"]],
+                "ranges": {
+                    col: (min(df[col]) * 1.25, max(df[col] * 1.25))
+                    for col in df.columns.values.tolist()
+                    if col not in ["X", "Year", "Subject"]
```
Comment (DSWallach) [inline tesla_ver/bubble_chart/bubble_chart.py:53]: What happens if there is a non-numeric column with a header different from these?

File: tesla_ver/bubble_chart/bubble_chart.py [hunk 2]
```diff
@@ -245,7 +195,7 @@ def play_pause_switch(n_clicks):
     @app.callback(
         Output("time-slider", "value"), [Input("play-interval", "n_intervals")], [State("time-slider", "value")]
     )
-    def play_increment(n_intervals, time_value):
+    def playIncrement(n_intervals, time_value):
```
Comment (DSWallach) [inline tesla_ver/bubble_chart/bubble_chart.py:198]: Function names should be underscored. Only classes use CamelCase

File: tesla_ver/bubble_chart/bubble_chart_layout.py [hunk 3]
```diff
@@ -10,7 +10,7 @@
     id="graph_div",
     children=[
         # This handles the upload of files
-        dcc.Upload(id="upload", children=html.Div(["Drag and Drop or ", html.A("Select Files")]), multiple=True,),
+        # dcc.Upload(id="upload", children=html.Div(["Drag and Drop or ", html.A("Select Files")]), multiple=True,),
```
Comment (DSWallach) [inline tesla_ver/bubble_chart/bubble_chart_layout.py:13]: If this isn't used remove it

File: tesla_ver/data_uploading/data_uploading.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+from dash.dependencies import Input, Output, State
+from dash.exceptions import PreventUpdate
+from base64 import b64decode
+from io import StringIO
+
+from tesla_ver.data_uploading.data_uploading_layout import LAYOUT
+from tesla_ver.redis_manager import redis_manager
+
+
+def generate_data_uploading(server):
```
Comment (DSWallach) [inline tesla_ver/data_uploading/data_uploading.py:18]: Add docstring

File: tesla_ver/data_uploading/data_uploading.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    @app.callback(
+        [
+            Output("data-table-div", "style"),
+            Output("data-table-visualization", "data"),
+            Output("data-table-visualization", "columns"),
+        ],
+        [Input("upload", "contents"), Input("upload", "filename")],
+    )
+    def upload_data_parsing(contents, filename):
```
Comment (DSWallach) [inline tesla_ver/data_uploading/data_uploading.py:30]: Docstring

File: tesla_ver/data_uploading/data_uploading.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        def upload_string_to_df(content):
```
Comment (DSWallach) [inline tesla_ver/data_uploading/data_uploading.py:31]: Is it necessary for this to be a local function? If not it should be defined elsewhere

File: tesla_ver/data_uploading/data_uploading.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        df = pd.DataFrame()
+        if None in [contents, filename]:
+            raise PreventUpdate
+        if len(contents) > 1:
+            parse_multiple_contents(contents, filename)
+        else:
+            df = upload_string_to_df(*contents)
+            df[df < 0] = 0
+            df[~df.isin([np.nan, np.inf, -np.inf]).any(1)]
+            df = df.fillna(0)
```
Comment (DSWallach) [inline tesla_ver/data_uploading/data_uploading.py:60]: Add comment on why your performing these dataframe operations

File: tesla_ver/data_uploading/data_uploading.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+                {"visibility": "visible"},
+                df.to_dict("records"),
+                [{"name": i, "id": i, "deletable": True, "selectable": True} for i in df.columns],
+            ]
+
+    @app.callback(
+        Output("data-table-visualization", "style_data_conditional"),
+        [Input("data-table-visualization", "selected_columns"), Input("data-table-visualization", "selected_rows"),],
+    )
+    def update_styles(selected_columns, selected_rows):
```
Comment (DSWallach) [inline tesla_ver/data_uploading/data_uploading.py:71]: Convert comment to docstring

File: tesla_ver/data_uploading/data_uploading.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    @app.callback(
+        Output("upload-confirmation", "style"),
+        [Input("redis-upload-button", "n_clicks")],
+        [
+            State("data-table-visualization", "data"),
+            State("data-table-visualization", "selected_columns"),
+            State("data-table-visualization", "selected_rows"),
+        ],
+    )
+    def push_to_redis(button_clicks, data_dict, selected_columns, selected_rows):
```
Comment (DSWallach) [inline tesla_ver/data_uploading/data_uploading.py:87]: docstring

File: tesla_ver/data_uploading/data_uploading.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        if None in [button_clicks, data_dict, selected_columns]:
+            raise PreventUpdate
```
Comment (DSWallach) [inline tesla_ver/data_uploading/data_uploading.py:89]: This type of thing seems to happen a lot. possibly t could be broken out as a function decorator

File: tesla_ver/data_uploading/data_uploading.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+            "data_numeric",
+            serialization_context.serialize(df[sorted(set(["Year", "Subject", *selected_columns]))])
+            .to_buffer()
+            .to_pybytes(),
+        )
+        redis_manager.redis.set(
+            "data_mdata",
+            serialization_context.serialize(df[["Year", "Subject", *mdata_cols]]).to_buffer().to_pybytes(),
+        )
+        redis_manager.redis.set("test", "Hello World, this is a test!")
```
Comment (DSWallach) [inline tesla_ver/data_uploading/data_uploading.py:111]: What is redis_manager?

File: tesla_ver/data_uploading/data_uploading.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+            State("data-table-visualization", "selected_rows"),
+        ],
+    )
+    def push_to_redis(button_clicks, data_dict, selected_columns, selected_rows):
+        if None in [button_clicks, data_dict, selected_columns]:
+            raise PreventUpdate
+
+        df = (
+            pd.DataFrame.from_records(data_dict)
+            if len(selected_rows) == 0
```
Comment (mstapylton) [inline tesla_ver/data_uploading/data_uploading.py:93]: I'd replace with `if not selected_rows:`

File: tesla_ver/data_uploading/data_uploading.py (hunk 7 above)
Comment (mstapylton) [reply tesla_ver/data_uploading/data_uploading.py:60]: Especially line 59, not really clear what that line is doing.
`any(1) `may be shorter, but `any(axis=1)` makes it more clear what's happening.

File: tesla_ver/data_uploading/data_uploading.py (hunk 12 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py]: addressed in 5a1d5c3

File: tesla_ver/data_uploading/data_uploading.py (hunk 9 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py:95]: addressed in 5a1d5c3

File: tesla_ver/data_uploading/data_uploading.py (hunk 8 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py:71]: addressed in 5a1d5c3

File: tesla_ver/bubble_chart/bubble_chart.py (hunk 2 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart/bubble_chart.py]: addressed in 5a1d5c3

File: tesla_ver/bubble_chart/bubble_chart_layout.py (hunk 3 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart/bubble_chart_layout.py]: addressed in 5a1d5c3

File: tesla_ver/data_uploading/data_uploading.py (hunk 4 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py:19]: addressed in 5a1d5c3

File: tesla_ver/data_uploading/data_uploading.py (hunk 5 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py:43]: addressed in 5a1d5c3

File: tesla_ver/data_uploading/data_uploading.py (hunk 7 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py:60]: addressed in 5a1d5c3

File: tesla_ver/data_uploading/data_uploading.py (hunk 6 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py]: moved to `/data_uploading/utils.py` 5a1d5c3

File: tesla_ver/data_uploading/data_uploading.py (hunk 11 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py]: `redis_manager` is a class that handles connecting to the redis data store once, and then can be imported anywhere to use the redis connection without connecting again every time.  It's located in `data_uploading/redis_manager.py`

File: tesla_ver/bubble_chart/bubble_chart.py (hunk 1 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart/bubble_chart.py:57]: Right now, those values are hardcoded, and the data files need to fit that pattern -- I'm not thinking of a good way besides either having the user rename the columns later, or some particularly large regex to try and guess at what the ID and time values are.

File: tesla_ver/data_uploading/data_uploading.py (hunk 10 above)
Comment (alexanderkyim) [reply tesla_ver/data_uploading/data_uploading.py:111]: Frankly I'm a little confused as to how to write a good decorator to do this -- I've PM'd you in slack to ask further about this.

**Comment by mstapylton** [comment]
1. I've set up docker and have the following questions and notes:
Build and launch the project with docker-compose up --build, 

Done 

2.
then go to the data uploading screen. Upload data there, select the identifying column, year column, and data columns to graph, as well as any rows that you wish to graph (if no rows are selected, all rows will be graphed)

Which is the identifying column? I selected tesla_data_10.tsv and saw a column name: subject...

Screenshot from 2020-07-27 10-54-47

I selected the subject, year and the data columns and then I then returned to the bubble chart page by clicking the back button in chrome (not good practice if that's the only way) and then going to bubble chart webpage.

Initially when I tried to select x-axis and y-axis values, nothing came up. Then when I clicked graph , the timeline values changed and I could select Bacteria 0, Bacteria 1 from the dropdowns:

Screenshot from 2020-07-27 10-56-26

But nothing was graphed. Let me know your thoughts on this, certainly the UI/UX will need improvements going forward and for a new feature like this, more detailed instructions would be helpful.

**Comment by alexanderkyim** [comment]
@mstapylton , I just pushed a fix for a few issues that I was running into on the new branch.  You shouldn't need to press graph again -- it's just to load the data, not to graph the values. Could you pull the latest changes and try again?

**Comment by DSWallach** [comment]
It looks like my comment never posted. When I try to run docker I get an error saying 'wsgi' can't be found. There doesn't seem to be a dependency for it in the environment file

**Comment by alexanderkyim** [comment]
@DSWallach , That's very strange, since wsgi is located at the root of the project and the dependency for it is werkzeug as far  as I know.

**Pull request merged**